diff -ur ../grpc-1.6.0/Makefile ./Makefile
--- ../grpc-1.6.0/Makefile	2017-08-29 19:24:15.000000000 +0000
+++ ./Makefile	2017-09-06 17:27:11.811775909 +0000
@@ -322,7 +322,7 @@
 HOST_LD ?= $(LD)
 HOST_LDXX ?= $(LDXX)
 
-CFLAGS += -std=c99 -Wsign-conversion -Wconversion $(W_SHADOW) $(W_EXTRA_SEMI)
+CFLAGS += -std=c99 -Wsign-conversion -Wconversion $(W_SHADOW) $(W_EXTRA_SEMI) -Wno-error=unused-function -Wno-error=deprecated-declarations -Wno-error=conversion
 CXXFLAGS += -std=c++11
 ifeq ($(SYSTEM),Darwin)
 CXXFLAGS += -stdlib=libc++
